sdk/tests/js/tests/tex-image-and-sub-image-2d-with-canvas.js

```diff
@@ -195,13 +324,41 @@ function generateTest(internalFormat, pixelFormat, pixelType, prologue, resource
                       },
                       debug);
             } else {
+                var localRed   = semiTransparent ? semiTransparentRedColor : redColor;
+                var localGreen = semiTransparent ? semiTransparentGreenColor : greenColor;
+
+                // Allow a tolerance for premultiplication/unmultiplication, especially for texture
+                // formats with lower bit depths.
+                var tolerance = 0;
+                if (semiTransparent) {
+                    tolerance = 3;
+                    if (pixelType == 'UNSIGNED_SHORT_5_6_5' || internalFormat == 'RGB565') {
+                        tolerance = 6;
+                    } else if (pixelType == 'UNSIGNED_SHORT_4_4_4_4' || internalFormat == 'RGBA4') {
+                        tolerance = 9;
+                    } else if (pixelType == 'UNSIGNED_SHORT_5_5_5_1' || internalFormat == 'RGB5_A1') {
+                        // It's ambiguous whether the one bit of alpha should be treated as:
```

Comment:
Thanks for finding that spec reference. Rewrote the comment to point to it. The opaque case is already tested; the semi-transparent case though can't be tested well with these canvas-to-texture tests, because most browsers use a premultiplied backing store for 2D canvases so nothing except transparent black is representable when alpha=0.

Filed #2454 about expanding another preexisting test to cover these cases.